lib/jnpr/junos/device.py

```diff
@@ -206,6 +206,119 @@ def port(self):
+                else:
+                    for re_state in self.facts['current_re']:
+                        # Multi-chassis case. A chassis master/backup, but
+                        # not the system master/backup.
+                        if '-backup' in re_state or '-master' in re_state:
+                            master = False
+                            break
+        except KeyError:
+            # Old style fact gathering is likely in use. Nothing we can do.
+            pass
+        return master
+
+    @master.setter
+    def master(self, value):
+        """ read-only property """
+        raise RuntimeError("master is read-only!")
+
+    # ------------------------------------------------------------------------
+    # property: re_name
+    # ------------------------------------------------------------------------
+
+    @property
+    def re_name(self):
+        """
+        The name of the current Routing Engine.
+
+        The current Routing Engine is the RE to which the NETCONF session is
+        connected.
+
+        Note: This property is based on new-style fact gathering. If old-style
+              fact gathering is in use, this property will return None.
+
+        :returns: A string containing the name of the current Routing Engine or
+                  None if unable to determine the state of the current
+                  Routing Engine.
+        """
+        re_name = None
+
+        try:
+            # Make sure the 'current_re' and 'hostname_info' facts have values
+            if (self.facts['current_re'] is not None and
+               self.facts['hostname_info'] is not None):
+                # re_name should be the intersection of the values in the
+                # 'current_re' fact and the keys in the 'hostname_info' fact.
+                intersect = (set(self.facts['current_re']) &
+                             set(self.facts['hostname_info'].keys()))
+                # intersect should usually contain a single element (the RE's
+                # name) if things worked correctly.
+                if len(intersect) == 1:
+                    re_name = list(intersect)[0]
+                # If intersect contains no elements
+                elif len(intersect) == 0:
+                    # Look for the first value
+                    # in 'current_re' which contains '-re'.
+                    for re_state in self.facts['current_re']:
+                        if '-re' in re_state:
+                            re_name = re_state
+                            break
+                    # Still haven't figured it out, if there's only one key
+                    # in 'hostname_info', assume that.
+                    if (re_name is None and
+                       len(self.facts['hostname_info']) == 1):
+                        re_name = self.facts['hostname_info'].keys()[0]
+        except KeyError:
+            # Old style fact gathering is likely in use. Nothing we can do.
+            pass
```

Comment:
I agree it will be a rare case, but one that I still think it's appropriate to handle.